File: src/segment-tree/lc3479.cc
#include <climits>
#include <vector>

using namespace std;

class Solution {
 public:
  // 线段树
  vector<int> segmentTree;
  vector<int> baskets;

  void build(int left, int right, int root) {
    if (left == right) {
      this->segmentTree[root] = this->baskets[left];
      return;
    }

    int mid = left + ((right - left) >> 1);
    this->build(left, mid, root * 2);
    this->build(mid + 1, right, root * 2 + 1);

    this->segmentTree[root] =
        max(this->segmentTree[root * 2], this->segmentTree[root * 2 + 1]);
  }

  int query(int left, int right, int queryLeft, int queryRight, int node) {
    if (queryLeft > right || queryRight < left) {
      return INT_MIN;
    }
    if (queryLeft <= left && right <= queryRight) {
      return this->segmentTree[node];
    }
    auto mid = left + ((right - left) >> 1);
    return max(
        this->query(left, mid, queryLeft, queryRight, node * 2),
        this->query(mid + 1, right, queryLeft, queryRight, node * 2 + 1));
  }

  void update(int left, int right, int pos, int val, int root) {
    if (left == right) {
      this->segmentTree[pos] = val;
      return;
    }
    auto mid = left + ((right - left) >> 1);
    if (pos <= mid) {
      this->update(left, mid, pos, val, root * 2);
    } else {
      this->update(mid + 1, right, pos, val, root * 2 + 1);
    }
    this->segmentTree[root] =
        max(this->segmentTree[root * 2], this->segmentTree[root * 2 + 1]);
  }

  int numOfUnplacedFruits(vector<int>& fruits, vector<int>& baskets) {
    if (baskets.size() == 0) {
      return fruits.size();
    }

    this->segmentTree = vector<int>(baskets.size() * 4);
    this->baskets = baskets;
    build(0, baskets.size() - 1, 1);
    auto ans = 0;

    for (auto i = 0; i < fruits.size(); i++) {
      auto left = 0;
      auto right = static_cast<int>(baskets.size() - 1);
      auto res = -1;
      while (left <= right) {
        auto mid = left + ((right - left) >> 1);
        if (this->query(0, baskets.size() - 1, 0, mid, 1) >= fruits[i]) {
          res = mid;
          right = mid - 1;
        } else {
          left = mid + 1;
        }
      }

      if (res != -1 && baskets[res] >= fruits[i]) {
        update(0, baskets.size() - 1, res, INT_MIN, 1);
      } else {
        ans++;
      }
    }
    return ans;
  }
};

File: src/segment-tree/lc3479.py
from typing import List


baseArr: List[int] = []
segmentTree: List[int] = []

NegInfinity = -(1 << 31)


# 对 baseArr 的 [left, right] 区间建立线段树, root 是根节点的下标
def build(left: int, right: int, root: int) -> None:
    if left == right:
        segmentTree[root] = baseArr[left]
        return

    mid = left + ((right - left) // 2)
    build(left, mid, root * 2)
    build(mid + 1, right, root * 2 + 1)

    segmentTree[root] = max(segmentTree[root * 2], segmentTree[root * 2 + 1])


def query(left: int, right: int, queryLeft: int, queryRight: int, node: int) -> int:
    if queryLeft > right or queryRight < left:
        return NegInfinity

    if queryLeft <= left and right <= queryRight:
        print(f"left: {left}, right: {right}")
        print(f"queryLeft: {queryLeft}, queryRight: {queryRight}")
        return segmentTree[node]

    mid = left + ((right - left) // 2)
    return max(
        query(left, mid, queryLeft, queryRight, node * 2),
        query(mid + 1, right, queryLeft, queryRight, node * 2 + 1),
    )


def update(left: int, right: int, idx: int, val: int, root: int) -> None:
    if left == right:
        segmentTree[root] = val
        return

    mid = left + ((right - left) // 2)
    if idx <= mid:
        update(left, mid, idx, val, root * 2)
    else:
        update(mid + 1, right, idx, val, root * 2 + 1)

    segmentTree[root] = max(segmentTree[root * 2], segmentTree[root * 2 + 1])


class Solution:
    def numOfUnplacedFruits(self, fruits: List[int], baskets: List[int]) -> int:
        if len(baskets) == 0:
            return len(fruits)

        global baseArr
        global segmentTree
        baseArr = baskets
        #! len(segmentTree) == 4 * len(baseArr)
        segmentTree = [0 for _ in range(len(baseArr) * 4)]

        # left, right, root
        build(0, len(baseArr) - 1, 1)
        ans = 0

        for i in range(len(fruits)):
            left = 0
            right = len(baseArr) - 1
            res = -1

            while left <= right:
                mid = left + ((right - left) // 2)
                if (
                    query(
                        left=0,
                        right=len(baseArr) - 1,
                        queryLeft=0,
                        queryRight=mid,
                        node=1,
                    )
                    >= fruits[i]
                ):
                    res = mid
                    right = mid - 1
                else:
                    left = mid + 1

            if res != -1 and baseArr[res] >= fruits[i]:
                update(left=0, right=len(baseArr) - 1, idx=res, val=NegInfinity, root=1)
            else:
                ans += 1

        return ans
